Ensure the overlay menu works when inserting navigation block patterns and that inner blocks are retained

talldan opened the pull request:
## Description
Fixes #36670
Fixes #37354

This fixes the two issues mentioned above. When testing the first one, I noticed the second one was no longer working.

The first issue is that the `UnsavedInnerBlocks` component that's shown when the block has standard uncontrolled inner blocks (this is hard to describe without using technical terms) didn't use the `ResponsiveWrapper` component that's used for showing the overlay menu. That was a pretty simple fix.

The second issue was introduced when another fix went in recently (https://github.com/WordPress/gutenberg/pull/36880). It's understandable as the `UnsavedInnerBlocks` behavior there isn't all that easy to follow, it exists to support an edge case, and also there are no end to end tests.

## How has this been tested?
1. Open the post editor
2. Switch to code editor mode
3. Paste in the following markup:
```html
<!-- wp:navigation {"overlayMenu":"always"} -->
<!-- wp:navigation-link {"label":"Hello world!","type":"post","id":1,"url":"http://localhost:8888/?p=1","kind":"post-type","isTopLevelLink":true} /-->
<!-- wp:navigation-link {"label":"Sample Page","type":"page","id":2,"url":"http://localhost:8888/?page_id=2","kind":"post-type","isTopLevelLink":true} /-->
<!-- /wp:navigation -->
```
4. Confirm that the overlay menu shows correctly and that the inner blocks are not lost.

Also test that the fix in #36880 continues to work.

## Types of changes
<!-- What types of changes does your code introduce?  -->
Bug fix (non-breaking change which fixes an issue)
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

talldan committed:
Show overlay menu for unsaved inner blocks too

talldan committed:
Fix uncontrolled inner blocks being deleted

talldan commented on packages/block-library/src/navigation/use-navigation-menu.js:
```diff
@@ -22,16 +22,10 @@ export default function useNavigationMenu( ref ) {
 				? getEditedEntityRecord( ...navigationMenuSingleArgs )
 				: null;
 
-			// getEditedEntityRecord will return the post regardless of status.
-			// Therefore if the found post is not published then we should ignore it.
-			if ( navigationMenu?.status !== 'publish' ) {
-				navigationMenu = null;
-			}
-
```
I removed this because the `wp_navigation` menus being created by `UnsavedInnerBlocks` were initially being created as drafts on purpose. Though it looks like that was also removed.

Not sure what to do now 🤔

talldan replied:
Switched it back to how it is in trunk

talldan committed:
Add e2e test for converting markup inner blocks to a navigation post

talldan committed:
Change useNavigationMenu back to trunk

talldan committed:
Re-enable single entity test

talldan commented on packages/e2e-tests/specs/editor/blocks/navigation.test.js:
```diff
@@ -564,6 +573,112 @@ describe( 'Navigation', () => {
+		expect( await getNavigationMenuRawContent() ).toMatchSnapshot();
+	} );
+
+	describe( 'Creating and restarting', () => {
```
This is the end-to-end test added in #36880. It was skipped, but I've fixed and re-enabled it here so that it asserts that the changes from that PR still work.

getdave replied:
Noted these don't provide full coverage for the issue uncovered in https://github.com/WordPress/gutenberg/pull/37358#pullrequestreview-832922134. Hopefully simple to add that coverage though?

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -142,25 +144,44 @@ function Navigation( {
+			// inner blocks, while `getBlocks` does. It might be more stable to
+			// introduce a selector like `getUncontrolledInnerBlocks`, just in
+			// case `getBlock` is fixed.
+			const _uncontrolledInnerBlocks = getBlock( clientId ).innerBlocks;
```
This doesn't seem to work. In my testing despite the entity being created and the code view of the markup showing no inner blocks, the block editor store wasn't in sync and thus still showed inner blocks.

Easy to check:

```
wp.data.select('core/block-editor').getBlocks()
```

This causes problems later on.

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -293,15 +314,17 @@ function Navigation( {
+	// clean slate.
 	useEffect( () => {
-		if ( ref === undefined && innerBlocks.length > 0 ) {
+		if (
+			! hasUncontrolledInnerBlocks &&
+			controlledInnerBlocks?.length > 0 &&
```
These new conditionals cause a regression when you're trying to start a new menu from an existing Menu that previously had uncontrolled inner blocks and has been converted to controlled blocks.

Again this is because the block editor store isn't in sync and shows outdated inner blocks.

getdave replied:
This will start working _if_ we implement a change which gets the block editor store in sync with the controlled block (i.e. inner blocks should be removed once we've converted to an entity and controlled inner blocks).

talldan replied:
Yeah, that makes sense, thanks for catching it I'll look into a fix 👍 

edit: oh, just noticed a fix was already pushed, thanks for doing that.

getdave reviewed (commented):
Thank you for working on these issues.

It fixes the Core issues you reference 🥳 

**However**...unfortunately I found this PR regresses https://github.com/WordPress/gutenberg/pull/36880.

- New Post
- Switch to code view and paste in:
```
<!-- wp:navigation {"overlayMenu":"none"} -->
<!-- wp:navigation-link {"label":"Hello world!","type":"post","id":1,"url":"http://localhost:8888/?p=1","kind":"post-type","isTopLevelLink":true} /-->

<!-- wp:navigation-link {"label":"Sample Page","type":"page","id":2,"url":"http://localhost:8888/?page_id=2","kind":"post-type","isTopLevelLink":true} /-->
<!-- /wp:navigation -->
```
- Update without switching out of code view.
- Switch to Visual Mode and select the Nav block - it will be converted to controlled blocks.
- Click `Select Menu -> Create New`
- Observe the broken UI state with an infinite spinner.
- Check `wp.data.select('core/block-editor').getBlocks()` and look for the Nav block. See that it still has inner blocks.

I have tried to push a fix which manually updates the innerblocks in the block editor store.

I'd prefer this fix was separated out from the overlay menu so it's easier to revert as required, but at this stage it's not a blocker.

getdave committed:
Attempt to sync block editor store with controlled state

Addresses https://github.com/WordPress/gutenberg/pull/37358#discussion_r769699131

getdave commented on packages/e2e-tests/specs/editor/blocks/navigation.test.js:
```diff
@@ -564,6 +573,112 @@ describe( 'Navigation', () => {
 		expect( quickInserter ).toBeTruthy();
 	} );
 
+	it( 'supports navigation blocks that have inner blocks within their markup and converts them to wp_navigation posts', async () => {
```
Can we add another test like this one where inner blocks are created in code view and then we allow them to be converted to an entity and _then_ try and Create a new Navigation using `Select Menu -> Create New`.

Prior to my fix that test should break. I think we need that coverage.

talldan committed:
Add e2e test for lingering uncontrolled blocks issue

talldan commented:
Added an e2e test for the fix, thanks for pushing a fix. As you say in the comment, blocks shouldn't be concerned about this, but I think this is the first use case of a core block that has both uncontrolled and controlled blocks.

getdave reviewed (approved):
Tested yesterday and only thing that changed since is addition of an e2e test. :shipit: 

Thanks again for fixing this bug.

bph commented:
Tested it on Wednesday and both issue cases seems to work now. Big Thank you. 🙂
